_pytest/assertion/rewrite.py

```diff
@@ -144,28 +145,25 @@ def _should_rewrite(self, name, fn_pypath, state):
         if fn_pypath.basename == 'conftest.py':
             state.trace("rewriting conftest file: %r" % (fn,))
             return True
-        elif self.session is not None:
+
+        if self.session is not None:
             if self.session.isinitpath(fn):
                 state.trace("matched test file (was specified on cmdline): %r" %
                             (fn,))
                 return True
-            else:
-                # modules not passed explicitly on the command line are only
-                # rewritten if they match the naming convention for test files
-                session = self.session  # avoid a cycle here
-                self.session = None
-                try:
-                    for pat in self.fnpats:
-                        if fn_pypath.fnmatch(pat):
-                            state.trace("matched test file %r" % (fn,))
-                            return True
-                finally:
-                    self.session = session
-                    del session
-        else:
-            for marked in self._must_rewrite:
-                if marked.startswith(name):
-                    return True
+
+        # modules not passed explicitly on the command line are only
+        # rewritten if they match the naming convention for test files
+        for pat in self.fnpats:
+            if fnmatch(fn_pypath.basename, pat):
+                state.trace("matched test file %r" % (fn,))
+                return True
+
+        for marked in self._must_rewrite:
+            if marked.startswith(name):
```

Comment:
The `_must_rewrite` set can also be full module paths, like `myplugin.plugin`.  This way round `myplugin.plugin.foo` or `myplugin.plugin.__init__` would not be rewritten but if it was the other way round it would be.  I guess it could just be a simple `==` comparison though?